File: Data Structures Intro/Hash Tables/C#/hash-table-task-net/HashTable/HashTable.cs
using System;
using System.Collections.Generic;

namespace HashTable
{
    /// <summary>
    /// Simple hash table implementation using chaining for collision resolution.
    /// </summary>
    /// <typeparam name="K">The type of the key</typeparam>
    /// <typeparam name="V">The type of the value</typeparam>
    public class HashTable<K, V>
    {
        private const int DefaultBucketCount = 100;
        private const int MinimumBucketCount = 1;

        /// <summary>
        /// Number of buckets in the hash table.
        /// </summary>
        private readonly int _bucketCount;

        /// <summary>
        /// Array of buckets for storing key-value pairs.
        /// </summary>
        private readonly Bucket<K, V>[] _buckets;

        /// <summary>
        /// Gets the buckets array. This is exposed for testing purposes only.
        /// </summary>
        public IReadOnlyList<Bucket<K, V>> Buckets => _buckets;

        /// <summary>
        /// Initializes a new instance of the HashTable class with the specified number of buckets.
        /// </summary>
        /// <param name="bucketCount">The number of buckets to create</param>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when bucketCount is less than 1.</exception>
        public HashTable(int bucketCount)
        {
            if (bucketCount < MinimumBucketCount)
                throw new ArgumentOutOfRangeException(nameof(bucketCount), $"Bucket count must be >= {MinimumBucketCount}");

            _bucketCount = bucketCount;
            _buckets = new Bucket<K, V>[_bucketCount];
            for (int i = 0; i < _bucketCount; i++)
                _buckets[i] = new Bucket<K, V>();
        }

        /// <summary>
        /// Initializes a new instance of the HashTable class with the default number of buckets.
        /// </summary>
        public HashTable() : this(DefaultBucketCount)
        {
        }

        /// <summary>
        /// Computes the hash value for the given key and maps it to a bucket index.
        /// </summary>
        private int GetBucketIndex(K key)
        {
            if (key == null)
                throw new ArgumentNullException(nameof(key));

            return Math.Abs(key.GetHashCode()) % _bucketCount;
        }

        /// <summary>
        /// Gets the bucket for the given key.
        /// </summary>
        public Bucket<K, V> GetBucket(K key)
        {
            return _buckets[GetBucketIndex(key)];
        }

        /// <summary>
        /// Adds or updates a key-value pair in the hash table.
        /// </summary>
        public void Set(K key, V value)
        {
            GetBucket(key).Put(key, value);
        }

        /// <summary>
        /// Retrieves the value associated with the given key.
        /// </summary>
        public V Get(K key)
        {
            return GetBucket(key).Get(key);
        }

        /// <summary>
        /// Removes the key-value pair associated with the given key.
        /// </summary>
        public bool Remove(K key)
        {
            return GetBucket(key).Remove(key);
        }

        /// <summary>
        /// Gets the total number of key-value pairs in the hash table.
        /// </summary>
        public int Count
        {
            get
            {
                int total = 0;
                foreach (var bucket in _buckets)
                    total += bucket.Count;
                return total;
            }
        }
    }
}

File: Data Structures Intro/Hash Tables/Java/hash-table-task-java/src/main/java/com/epam/bsp/hashtable/HashTable.java
package com.epam.bsp.hashtable;

/**
 * Simple hash table implementation
 * @param <K> type of key
 * @param <V> type of value
 */
public class HashTable<K, V> {

    /**
     * Number of buckets
     */
    private int nBuckets;
    /**
     * Array of buckets
     */
    private Bucket<K, V>[] buckets;

    @SuppressWarnings("unchecked")
    public HashTable(int nBuckets) {
        this.nBuckets = nBuckets;
        this.buckets = (Bucket<K, V>[]) new Bucket[nBuckets];
        for (int i = 0; i < nBuckets; i++) {
            buckets[i] = new Bucket<>();
        }
    }

    public HashTable() {
        this(100);
    }

    public Bucket<K, V>[] getBuckets() {
        return buckets;
    }

    private int h(K key) {
        return (key.hashCode() % nBuckets + nBuckets) % nBuckets;
    }

    public Bucket<K, V> getBucket(K key) {
        return buckets[h(key)];
    }

    /**
     * Inserts or replaces a (K,V) pair.
     */
    public void set(K key, V value) {
        getBucket(key).put(key, value);
    }

    /**
     * Retrieves a value by key.
     */
    public V get(K key) {
        return getBucket(key).get(key);
    }

    /**
     * Removes a (K,V) pair.
     */
    public boolean remove(K key) {
        return getBucket(key).remove(key);
    }
}
